### ISSUE TITLE
Errors after latest Flutter update - Breaking chages with FocusScopeNode

### ISSUE THREAD

**Issue by Skquark** [memspace/zefyr#109]
After a Flutter update to latest Dev Channel, getting a bunch of compile errors that have not been reported yet.  Looks like they made some changes to the way we handle the FocusNode that shouldn't be that difficult to fix.  Here are the errors:
```
file:///d:/Program%20Files/Android/flutter/.pub-cache/hosted/pub.dartlang.org/zefyr-0.5.0/lib/src/widgets/editable_text.dart:102:28: Error: The method 'reparentIfNeeded' isn't defined for the class 'FocusScopeNode'.
 - 'FocusScopeNode' is from 'package:flutter/src/widgets/focus_manager.dart' ('file:///d:/Program%20Files/Android/flutter/packages/flutter/lib/src/widgets/focus_manager.dart'). 
Try correcting the name to the name of an existing method, or defining a method named 'reparentIfNeeded'.
    FocusScope.of(context).reparentIfNeeded(focusNode);
                           ^^^^^^^^^^^^^^^^
file:///d:/Program%20Files/Android/flutter/.pub-cache/hosted/pub.dartlang.org/zefyr-0.5.0/lib/src/widgets/selection.dart:483:22: Error: Too few positional arguments: 5 required, 4 given.
        .buildToolbar(context, editingRegion, midpoint, widget.delegate);
                     ^
file:///d:/Program%20Files/Android/flutter/.pub-cache/hosted/pub.dartlang.org/zefyr-0.5.0/lib/src/widgets/input.dart:153:18: Error: The argument type 'String' can't be assigned 
to the parameter type 'DiagnosticsNode'.
 - 'DiagnosticsNode' is from 'package:flutter/src/foundation/diagnostics.dart' ('file:///d:/Program%20Files/Android/flutter/packages/flutter/lib/src/foundation/diagnostics.dart').
Try changing the type of the parameter, or casting the argument to 'DiagnosticsNode'.
        context: 'while updating editing value',
                 ^
```
I investigated it a bit and found reparentIfNeeded and detatch is obsolete now and needs to be reimplemented.  Here's a useful discussion about the change: [https://github.com/flutter/flutter/issues/31784](https://github.com/flutter/flutter/issues/31784)
So just playing around, here's my quick fix in editable_text.dart:
```
    //FocusScope.of(context).reparentIfNeeded(focusNode);
    FocusScopeNode _node = focusNode;
    FocusAttachment _nodeAttachment = _node.attach(context);
    _nodeAttachment.reparent();
```
Although FocusScopeNode and FocusAttachment should be declared in the class, not in the build function where I put it for testing, so that's not the right way to do it but at least got rid of one of the errors.  A similar incompatibility affected other package I use called Flushbar, and here was their changes to address the issue: [https://github.com/AndreHaueisen/flushbar/compare/dev_version](https://github.com/AndreHaueisen/flushbar/compare/dev_version) which could be a helpful example in refactoring.
Regarding the breaking change with DiagnosticNode in input.dart, looks like an easy change to make the line in the FlutterErrorDetails from context: 'while updating editing value', into context: ErrorDescription('while updating editing value'), which was covered in this announcement [https://groups.google.com/forum/#!topic/flutter-announce/hp1RNIgej38](https://groups.google.com/forum/#!topic/flutter-announce/hp1RNIgej38)
Hope this helps and we can see an updated release soon, or at least a PR fork so we can get back to building our projects again.  Thanks.

**Comment by pulyaevskiy** [comment]
Thanks for a thorough description! 👍

Will take a look as soon as I can. This does sound like a straightforward fix.

**Comment by Skquark** [comment]
No problem, except for now there's a split between people on the Stable channel and those on Dev or Master, so you might need two versions to accommodate until things get merged to Stable.  A suggested workaround at least on the DiagnosticNode error is to make that line look like this:
context: (FlutterError('text') is Diagnosticable) ? ErrorDescription('while updating editing value') : 'while updating editing value' ,
but that doesn't help the FocusScopeNode changes that won't work for people on Stable yet.

Regarding the other error with the buildToolbar arguments in selection.dart line 482, looks like they added another parameter for List<TextSelectionPoint> endpoints after the Offset.  Looking at the code, it wasn't obvious to me what you would use there after the midpoint, but I suspect it would be similar to how you constructed the Offset midpoint only without dividing by 2.. Here's my guess for how you'd fix that:
```
     final Offset endpoint = new Offset(
      (boxes.length == 1)
          ? (boxes[0].start + boxes[0].end)
          : (boxes[0].start + boxes[1].start),
      boxes[0].bottom - block.preferredLineHeight,
    );
    final TextSelectionPoint textEndpoint = new TextSelectionPoint(endpoint, TextDirection.ltr);
    final toolbar = widget.controls
        .buildToolbar(context, editingRegion, midpoint, [textEndpoint], widget.delegate);
```
However I might be missing some nuances of how you'd do it more correctly.  Hope this speeds things along a bit.  Gotta love them breaking changes to shake things up...

**Comment by Skquark** [comment]
I'm still trying to work this out.  At least I'm not getting compile errors, but still got errors when using.  In my tests, I didn't go the FocusScopeNode reparenting correctly, so tried with this code instead:
```
    //FocusScope.of(context).reparentIfNeeded(focusNode);
    FocusAttachment _nodeAttachment = FocusScope.of(context).attach(context);
    _nodeAttachment.reparent(parent: focusNode);
```
However, when I tap in the editor, I get this error now:
```
I/flutter (15536): The following assertion was thrown while dispatching notifications for ZefyrController:
I/flutter (15536): The supplied child is already an ancestor of this node. Loops are not allowed.
I/flutter (15536): 'package:flutter/src/widgets/focus_manager.dart':
I/flutter (15536): Failed assertion: line 633 pos 12: '!ancestors.contains(child)'
I/flutter (15536):
I/flutter (15536): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter (15536): more information in this error message to help you determine and fix the underlying cause.
I/flutter (15536): In either case, please report this assertion by filing a bug on GitHub:
I/flutter (15536):   https://github.com/flutter/flutter/issues/new?template=BUG.md
I/flutter (15536):
I/flutter (15536): When the exception was thrown, this was the stack:
I/flutter (15536): #2      FocusNode._reparent 
package:flutter/…/widgets/focus_manager.dart:633
I/flutter (15536): #3      FocusNode.requestFocus 
package:flutter/…/widgets/focus_manager.dart:705
I/flutter (15536): #4      _ZefyrEditableTextState.requestKeyboard 
package:zefyr/…/widgets/editable_text.dart:82
I/flutter (15536): #5      _ZefyrEditableTextState._handleLocalValueChange 
package:zefyr/…/widgets/editable_text.dart:257
```
So I guess it's attaching to the wrong parent or something, not entirely sure the best way to set it up, but hopefully you guys have a better handle on it.. Maybe it doesn't need the FocusAttachment or needing the reparent, but I'm missing something in the architecture of it.

**Comment by pulyaevskiy** [comment]
Thanks everyone for all the details! 

I've just pushed an update to `dev` branch which should work with latest `dev` channel in Flutter repo.

You can install it with following in your `pubspec.yaml`:

```yaml
dependencies:
  zefyr:
    git:
      url: https://github.com/memspace/zefyr.git
      ref: dev
```

Please let me know if you notice any regressions!

Some details on breaking changes to focus nodes can be found here: https://github.com/flutter/flutter/issues/31784#issuecomment-487678661

**Comment by w4mxl** [comment]
@pulyaevskiy 

After install it as : 

```
dependencies:
  zefyr:
    git:
      url: https://github.com/memspace/zefyr.git
      ref: dev
```

I got "Could not find a file named "pubspec.yaml" in https://github.com/memspace/zefyr.git 56def8e0f4bf7af25baaa4ed52f7b22c1fee3db9."

Why is that?

Thanks.

**Comment by w4mxl** [comment]
After added path，it's ok now！

  ```
zefyr:
    git:
      url: https://github.com/memspace/zefyr.git
      ref: dev
      path: packages/zefyr
```

**Comment by Skquark** [comment]
Yup, was working for a bit, but after I did the latest flutter upgrade, I'm getting all those new errors too..  Looked into it and found it discussed here: https://github.com/flutter/flutter/pull/32059 as breaking change.. As an experiment, I modified my selection.dart and added the simple suggestions into _ZefyrSelectionOverlayState which extends TextSelectionDelegate:
```
  bool get copyEnabled => true;
  bool get cutEnabled => true;
  bool get pasteEnabled => true;
  bool get selectAllEnabled => true;
```
and it actually worked!  Seems too easy, but if that's the proper fix, might want to add that to the dev release...

**Comment by Skquark** [comment]
Now there's another related break with my last Flutter update, they're makin a lot of changes up there.  Here's the latest Zefyr error:
```
file:///D:/Program%20Files/Android/flutter/.pub-cache/git/zefyr-56def8e0f4bf7af25baaa4ed52f7b22c1fee3db9/packages/zefyr/lib/src/widgets/rich_text.dart:184:26: Error: The parameter 'value' of the method 'RenderZefyrParagraph.text' has type 'TextSpan', which does not match the corresponding type, 'InlineSpan', in the overridden method, 'RenderParagraph.text'.
 - 'TextSpan' is from 'package:flutter/src/painting/text_span.dart' ('file:///D:/Program%20Files/Android/flutter/packages/flutter/lib/src/painting/text_span.dart').
  set text(InlineSpan value) {
      ^
Compiler terminated unexpectedly.
```
So I dug into rich_text.dart to this line:
```
  void set text(TextSpan value) {
    _prototypePainter.text = new TextSpan(text: '.', style: value.style);
    _selectionRects = null;
    super.text = value;
  }
```
Searched the Flutter issues forum and found a few breaking changes with TextSpan like this 
[https://github.com/flutter/flutter/pull/33794](https://github.com/flutter/flutter/pull/33794) and [https://github.com/flutter/flutter/pull/33946](https://github.com/flutter/flutter/pull/33946)
I got bored of reading about it so I just replaced TextSpan with InlineSpan on line 184 and it worked .... Not sure if there's any other impact that I'm not looking at, but at least it compiles.

**Comment by pulyaevskiy** [comment]
>  I just replaced TextSpan with InlineSpan

Yes, that's about it. I just pushed an update to `dev` branch with this change.

Note that I'm seeing strange issue with selection handles not being aligned properly with latest `dev` version of Flutter. Will look into this later.

**Comment by Luppakorva** [comment]
I'm struggling with this with the latest flutter update as well, I did change to dev but after that I'm getting: 

```
.pub-cache/git/zefyr-077cd10dfb04d05ae97801b49160f7af0829ca47/packages/zefyr/lib/src/widgets/selection.dart:509:49: Error: Too few positional arguments: 6 required, 5 given.
    final toolbar = widget.controls.buildToolbar(
```

Any hotfix for this or something I could try?

**Comment by pulyaevskiy** [comment]
All of these changes are now on master and was just released in 0.6.0. Closing as resolved.
